File: kt/Minsk/minsk/src/main/kotlin/dev/phytolizer/minsk/analysis/syntax/Lexer.kt
package dev.phytolizer.minsk.analysis.syntax

class Lexer(private val _text: String) : Iterable<SyntaxToken> {
    private val _diagnostics = mutableListOf<String>()
    val diagnostics: List<String>
        get() = _diagnostics

    override fun iterator(): Iterator<SyntaxToken> {
        return LexerIterator(this)
    }

    private class LexerIterator(private val lexer: Lexer) : Iterator<SyntaxToken> {
        private var position = 0
        private fun peek(offset: Int): Char {
            val index = position + offset
            return if (index >= lexer._text.length) {
                '\u0000'
            } else {
                lexer._text[index]
            }
        }

        private val current: Char
            get() = peek(0)

        override fun hasNext(): Boolean {
            return position < lexer._text.length
        }

        override fun next(): SyntaxToken {
            val start = position
            var currentText: String? = null
            var value: Any? = null
            var kind = SyntaxKind.BadToken

            if (current.isWhitespace()) {
                while (current.isWhitespace()) {
                    position += 1
                }

                kind = SyntaxKind.WhitespaceToken
            } else if (current.isDigit()) {
                while (current.isDigit()) {
                    position += 1
                }

                kind = SyntaxKind.NumberToken
                currentText = lexer._text.substring(start until position)
                try {
                    value = currentText.toInt()
                } catch (_: NumberFormatException) {
                    lexer._diagnostics.add("The number $currentText doesn't fit in an Int")
                }
            } else if (current.isLetter()) {
                while (current.isLetterOrDigit()) {
                    position += 1
                }

                currentText = lexer._text.substring(start until position)
                kind = SyntaxFacts.keywordKind(currentText)
            } else when (current) {
                '\u0000' -> {
                    kind = SyntaxKind.EndOfFileToken
                }
                '+' -> {
                    kind = SyntaxKind.PlusToken
                    position += 1
                }
                '-' -> {
                    kind = SyntaxKind.MinusToken
                    position += 1
                }
                '*' -> {
                    kind = SyntaxKind.StarToken
                    position += 1
                }
                '/' -> {
                    kind = SyntaxKind.SlashToken
                    position += 1
                }
                '(' -> {
                    kind = SyntaxKind.OpenParenthesisToken
                    position += 1
                }
                ')' -> {
                    kind = SyntaxKind.CloseParenthesisToken
                    position += 1
                }
                '!' -> if (peek(1) == '=') {
                    kind = SyntaxKind.BangEqualsToken
                    position += 2
                } else {
                    kind = SyntaxKind.BangToken
                    position += 1
                }
                '&' -> if (peek(1) == '&') {
                    kind = SyntaxKind.AmpersandAmpersandToken
                    position += 2
                }
                '|' -> if (peek(1) == '|') {
                    kind = SyntaxKind.PipePipeToken
                    position += 2
                }
                '=' -> if (peek(1) == '=') {
                    kind = SyntaxKind.EqualsEqualsToken
                    position += 2
                }
                else -> {}
            }

            if (kind == SyntaxKind.BadToken) {
                lexer._diagnostics.add("Unexpected character '$current'")
                position += 1
            }

            return SyntaxToken(kind, start, currentText ?: lexer._text.substring(start until position), value)
        }
    }
}

File: py/minsk/analysis/syntax/lexer.py
from typing import Any, Optional

from minsk.analysis.syntax.kind import SyntaxKind
from minsk.analysis.syntax.token import SyntaxToken


class Lexer:
    class _Iter:
        _text: str
        _position: int

        def __init__(self, text):
            self._text = text
            self._position = 0

        @property
        def _current(self) -> str:
            return self._peek(0)

        def __next__(self):
            if self._position >= len(self._text):
                raise StopIteration

            start = self._position
            kind = SyntaxKind.BadToken
            value: Optional[Any] = None
            current_text: Optional[str] = None

            match self._current:
                case c if c.isspace():
                    while self._current.isspace():
                        self._position += 1
                    kind = SyntaxKind.WhitespaceToken
                case c if c.isdigit():
                    while self._current.isdigit():
                        self._position += 1
                    kind = SyntaxKind.NumberToken
                    current_text = self._current_text(start)
                    try:
                        value = int(current_text)
                    except ValueError:
                        pass
                case "+":
                    kind = SyntaxKind.PlusToken
                    self._position += 1
                case "-":
                    kind = SyntaxKind.MinusToken
                    self._position += 1
                case "*":
                    kind = SyntaxKind.StarToken
                    self._position += 1
                case "/":
                    kind = SyntaxKind.SlashToken
                    self._position += 1
                case "(":
                    kind = SyntaxKind.OpenParenthesisToken
                    self._position += 1
                case ")":
                    kind = SyntaxKind.CloseParenthesisToken
                    self._position += 1

            if kind == SyntaxKind.BadToken:
                self._position += 1

            if current_text is None:
                current_text = self._current_text(start)

            return SyntaxToken(kind, start, current_text, value)

        def _peek(self, offset: int) -> str:
            index = self._position + offset
            if index >= len(self._text):
                return "\0"
            return self._text[index]

        def _current_text(self, start: int) -> str:
            return self._text[start:self._position]

    _text: str

    def __init__(self, text: str):
        self._text = text

    def __iter__(self):
        return Lexer._Iter(self._text)
